### PR TITLE
fixed contact: applied only when objects in contact close enough

### PR DESCRIPTION
**Pull request by vincentberenz**
## issue 
Contacts between the ball and the table/the racket have issues, e.g. the ball "falls" inside the table a bit or even goes through it

## reason
- the contact ball controller runs for each integration step, while it should run only at each mujoco step
- Mujoco starts reporting contacts early on, i.e. a few iterations before they occur. But the custom contact model assumes that the contact is occuring at the current step, which results in a negative duration used for computing the ball position following the contact

## solution
- code is refactored so that the contact ball controller computes contacts at each mujoco step only. Results a simplification of the code, e..g "contact_logic.hpp" is removed, and headers are rearranged accordingly
- the custom contact models returns "false" if there is a negative duration, and the contact ball controller considers then the contact can be ignored

## misc
The gap for the contact with the table in the mujoco xml file has been removed, as it is not used

### CODE REVIEW

File: python/pam_mujoco/models.py [hunk 1]
```diff
@@ -165,7 +165,7 @@ def defaults_solrefs():
 
 def defaults_gaps():
 
-    return {"ball": {"floor": 0.0, "table": 0.02}}
+    return {"ball": {"floor": 0.0, "table": 0.0}}
```
Comment (dtrb) [inline python/pam_mujoco/models.py:168]: Do we need to set table height back to .02? @simonguist: Why did we lower the table in the first place?

**Review by dtrb — APPROVED** [review]
lgtm but consider lowering the table again

File: python/pam_mujoco/models.py (hunk 1 above)
Comment (vincentberenz) [reply python/pam_mujoco/models.py:168]: This value is for tuning mujoco such that it ignores certain contacts. Looking at it now, I think we should use it, but I am still not sure how

Comment (simonguist) [reply python/pam_mujoco/models.py:168]: If I remember correctly the gap was mainly for the optics

**Review by simonguist — APPROVED** [review]

File: python/pam_mujoco/models.py (hunk 1 above)
Comment (vincentberenz) [reply python/pam_mujoco/models.py:168]: it has something to do with contact detection:
http://www.mujoco.org/book/XMLreference.html

**Pull request merged**